src/audio/volume/volume_hifi3.c

```diff
@@ -192,19 +242,40 @@ static void vol_s16_to_s16(struct comp_dev *dev, struct audio_stream *sink,
 			mult = AE_MULF32X16_L0(volume, in_sample);
 
 			/* Multiply of Q1.31 x Q1.15 gives Q1.47. Multiply of
-			 * Q8.16 x Q1.15 gives Q8.32, so need to shift left
-			 * by 31 to get Q1.63. Sample is Q1.31.
+			 * Q8.16 x Q1.15 (IPC3) or Q1.23 x Q1.15 (IPC4) gives
+			 * appropriately Q8.32 (IPC3) or Q1.39 (IPC4),
+			 * so need to shift left by 31 (IPC3) or 24 (IPC) to get
+			 * Q1.63. Sample is Q1.31.
 			 */
+#if CONFIG_IPC_MAJOR_3
 			out_sample = AE_ROUND32F64SSYM(AE_SLAI64S(mult, 31));
+#else /* CONFIG_IPC_MAJOR_4 */
+			out_sample = AE_ROUND32F64SSYM(AE_SLAI64S(mult, 24));
+#endif
 
 			/* Set sink as circular buffer */
 			vol_setup_circular(sink);
 
 			/* Round to Q1.15 and store the output sample */
 			AE_S16_0_XC(AE_ROUND16X4F32SSYM(out_sample, out_sample),
 				    out, sizeof(ae_int16));
+
+#if CONFIG_IPC_MAJOR_4
+			out_converted_sample =
+				(int32_t)AE_MOVINT32X2_FROMF32((ae_f32)out_sample);
+
+			if (out_converted_sample >
+			   cd->peak_regs.peak_meter_[channel])
+				cd->peak_regs.peak_meter_[channel] =
+					out_converted_sample;
+#endif
 		}
 	}
+#if CONFIG_IPC_MAJOR_4
+	/* update peakvol in mailbox */
+	mailbox_sw_regs_write(cd->mailbox_offset, &cd->peak_regs,
+			      sizeof(cd->peak_regs));
+#endif
```

Comment:
instead of calling this inside every volume conversion function, can we move this to outside of the call to cd->scale_vol() in volume.c?

Reply:
yes, we can but changed to peak_vol_calc&update as Liam proposed so it will stay in volume_hifi3.c